mm2src/common/jsonrpc_client.rs

```diff
@@ -9,41 +11,73 @@ use std::fmt;
 #[macro_export]
 macro_rules! rpc_func {
     ($selff:ident, $method:expr $(, $arg_name:expr)*) => {{
-        let mut params = vec![];
-        $(
-            params.push(json::value::to_value($arg_name).unwrap());
-        )*
-        let request = JsonRpcRequest {
-            jsonrpc: $selff.version().into(),
-            id: $selff.next_id(),
-            method: $method.into(),
-            params
-        };
+        let request = $crate::rpc_req!($selff, $method $(, $arg_name)*);
         $selff.send_request(request)
     }}
 }
 
 /// Macro generating functions for RPC requests.
-/// Send the RPC request to specified remote endpoint using the passed address.
+/// Sends the RPC request to specified remote endpoint using the passed address.
 /// Args must implement/derive Serialize trait.
 /// Generates params vector from input args, builds the request and sends it.
 #[macro_export]
 macro_rules! rpc_func_from {
-    ($selff:ident, $address:expr, $method:expr $(, $arg_name:ident)*) => {{
+    ($selff:ident, $address:expr, $method:expr $(, $arg_name:expr)*) => {{
+        let request = $crate::rpc_req!($selff, $method $(, $arg_name)*);
+        $selff.send_request_to($address, request)
+    }}
+}
+
+/// Macro generating functions for RPC requests.
+/// Args must implement/derive Serialize trait.
+/// Generates params vector from input args, builds the `JsonRpcRequest` request.
+#[macro_export]
+macro_rules! rpc_req {
+    ($selff:ident, $method:expr $(, $arg_name:expr)*) => {{
         let mut params = vec![];
         $(
             params.push(json::value::to_value($arg_name).unwrap());
         )*
-        let request = JsonRpcRequest {
+        JsonRpcRequest {
             jsonrpc: $selff.version().into(),
             id: $selff.next_id(),
             method: $method.into(),
             params
-        };
-        $selff.send_request_to($address, request)
+        }
     }}
 }
 
+pub type JsonRpcResponseFut =
+    Box<dyn Future<Item = (JsonRpcRemoteAddr, JsonRpcResponseEnum), Error = String> + Send + 'static>;
+pub type RpcRes<T> = Box<dyn Future<Item = T, Error = JsonRpcError> + Send + 'static>;
+
+pub trait BatchRequest {
+    /// Sends the RPC batch request.
+    /// Responses are guaranteed to be in the same order in which they were requested.
+    fn batch_rpc<Rpc, T>(self, rpc_client: &Rpc) -> RpcRes<Vec<T>>
+    where
+        Rpc: JsonRpcClient,
+        T: DeserializeOwned + Send + 'static;
+}
+
+impl<I> BatchRequest for I
```

Comment:
Yes, this makes sense! I wanted to use the trait this way:
```rust
tx_ids
                .iter()
                .map(|txid| rpc_req!(self, "getrawtransaction", txid, verbose))
                .batch_rpc(self)
```
But I agree that it might be confusing.

Reply:
Done